### PR TITLE
feedback auth result via lmq rpc

### LINKED ISSUE #1497: lokinet exit auth invalidates session before competing auth
**Describe the bug**
lokinet exit with auth resets session before auth can complete

**To Reproduce**
turn on exit with an auth code with lots of traffic on your end

**Screenshots Or Logs**
none

### PR DESCRIPTION
**Pull request by majestrate**
fixes #1497

### CODE REVIEW

File: llarp/rpc/endpoint_rpc.cpp [hunk 1]
```diff
@@ -33,15 +33,27 @@ namespace llarp::rpc
...
       std::function<void(service::AuthResult)> hook)
   {
+    service::ConvoTag tag = msg->tag;
+    m_PendingAuths.insert(tag);
     const auto from = msg->sender.Addr();
-    auto reply = [logic = m_Endpoint->RouterLogic(), hook](service::AuthResult result) {
-      logic->Call([hook, result]() { hook(result); });
-    };
+    auto reply =
+        [self = this, tag, logic = m_Endpoint->RouterLogic(), hook](service::AuthResult result) {
```
Comment (jagerman) [inline llarp/rpc/endpoint_rpc.cpp:51]: Why capture `self=this` rather than just `this`?

Comment (majestrate) [reply llarp/rpc/endpoint_rpc.cpp:51]: a horrid decrepit bad habit of mine.

File: llarp/service/auth.cpp [hunk 2]
```diff
@@ -18,6 +18,24 @@ namespace llarp::service
...
+      case AuthResult::eAuthRejected:
+        return "rejected";
+      case AuthResult::eAuthPaymentRequired:
+        return "payment required";
+      case AuthResult::eAuthRateLimit:
+        return "rate limit hit";
+      default:
+        return "unknown";
+    }
+  }
```
Comment (jagerman) [inline llarp/service/auth.cpp]: I'm thinking that there needs to be an "other" option here which could be returned with a short explanatory message.  I can imagine a bunch of random things a paid exit might want to return, e.g. "daily bandwidth exceeded" or "account suspended" or ...

Comment (jagerman) [reply llarp/service/auth.cpp]: Or perhaps: AuthResult has Accepted/Rejected, with an optional reason string that could carry the reason info.

**Comment by majestrate** [comment]
rebased on dev

File: llarp/service/auth.cpp (hunk 2 above)
Comment (majestrate) [reply llarp/service/auth.cpp]: added reason info.

File: llarp/service/outbound_context.cpp [hunk 3]
```diff
@@ -528,7 +528,90 @@ namespace llarp
...
+
+        m_Endpoint->RemoveConvoTag(frame.T);
+        if (authResultListener)
+        {
+          switch (frame.R)
+          {
+            case 1:
+              result.code = eAuthRejected;
+              break;
+            case 3:
```
Comment (tewinget) [inline llarp/service/outbound_context.cpp]: could use a comment noting why `case 2` is (apparently) omitted.

File: llarp/service/tag.hpp [hunk 4]
```diff
@@ -6,10 +6,7 @@
 #include <util/status.hpp>
 
 #include <sodium/crypto_generichash.h>
-
-#ifndef WIN32_LEAN_AND_MEAN
-#define WIN32_LEAN_AND_MEAN
-#endif
+#include <functional>
```
Comment (tewinget) [inline llarp/service/tag.hpp]: only change to this file, so likely unnecessary new include

**Review by tewinget — APPROVED** [review]
Couple of small notes, but looks good to me.

File: llarp/service/outbound_context.cpp (hunk 3 above)
Comment (jagerman) [reply llarp/service/outbound_context.cpp]: Or better: change them to be constexpr values rather than magic numbers.

Comment (majestrate) [reply llarp/service/outbound_context.cpp]: now uses an enum class

File: llarp/service/tag.hpp (hunk 4 above)
Comment (majestrate) [reply llarp/service/tag.hpp]: done

**Pull request merged**